README.md

Seqdiag.js
----------
sequence diagram on javascript from simple text format.

Example
------------

	<!doctype html>
	<head>
		<script type="text/javascript" src="../lib/seqdiag/parser.js" ></script>
		<script type="text/javascript" src="../lib/seqdiag.js" ></script>
	</head>
	<body>
		<figure id="diag"></figure>
		<script type="text/x-seqdiag" id="seqdiag-text">
			seqdiag {
				A [label = "webapp"];
				B [label = "DB"];

				A -> B[label = "sql"];
				B -> A [label = "result"];
			};
		</script>
		<script type="text/javascript">
			window.onload = function() {
				var taget = document.getElementById('seqdiag-text');
				var ast = SeqdiagParser.parse(target.innerText);
				var diagram = Seqdiag.Builder.build(ast);
				var svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
				svg.setAttribute("width", "1024");
				svg.setAttribute("width", "600");
				var drawer = new Seqdiag.Drawer.SVG(diagram, svg, document);
				drawer.draw();
				var figure = document.getElementById('diag');
				figure.appendChild(svg);
			};
		</script>
	</body>
	</html>

SEE ALSO
-----------
 * examples/try.html
 * seqdiag - http://blockdiag.com/en/seqdiag/

LICENSE
----------------
(c) 2012 Keiji Yoshimi

This project is released under MIT license.
See followings: http://www.opensource.org/licenses/MIT



CHANGES.md
Changes
===================

v0.0.3				2012-09-18 09:45:06 +0900
------------------------------------------------------------------------
 - refs #4 drawSelfReferenceEdge.
 - refs #4 build up self_reference edge.
 - refs #1 adjust edgeMargin.
 - refs #1 fix node's height.
 - adjust arrowWidth / arrowHeight
 - refactoring using isXXX attributes.
 - using raw type's value is not useful.
 - move out drawEdgeArrowHead for drawEdge's readablity.
 - move out drawEdgeText() for drawEdge()'s readability.
 - refs #3 draw async arrow.
 - type return edge should invert from and to.
 - edge's type name "asynchronous" is too long. So use "async".
 - test edge's type.
 - support comment statement.
 - support dotted to draw dotted.
 - added Edge to type.
 - support under_score as identifier.
 - support no defining node statement case.
 - note about license.
 - don't use magic number.
 - change magic number to class variable.
 - move doc to out of class.
 - added separator for source code readability.

v0.0.2				2012-09-12 10:21:24 +0900
------------------------------------------------------------------------
 - it should be half char height.
 - node's text position should be considered char height.
 - charWidth should be configuable.
 - adjust node's text position for centerize.
 - consider left margin.
 - changable color theme.
 - allow edge that have not attributes.

v0.0.1				2012-09-11 22:58:08 +0900
------------------------------------------------------------------------
 - initial release.


test/parser.js
require('./test_helper.js');
(function() {
    "use strict";
    var q = QUnit;

    q.module("SeqdiagParser", {
    });

    q.test('normal case', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" + 
            "A[label = \"aaa\", foo = \"bar\"];\n" +
            "B[label = \"あいうえお\"];\n" +
            "C;\n" +
            "A -> B [ label = \"A to B\" ];" +
            "B --> A;" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.ok(ast[1] instanceof Object, "graph has token attribute");
        q.equal(ast[1]["id"], "seqdiag", "graph attribute should have 'id'");

        q.ok(ast[1]["stmt"] instanceof Array, "graph attribute should have stmt");

        // for node A
        q.equal(ast[1]["stmt"][0][0], "node", "first statement should be node");
        q.equal(typeof ast[1]["stmt"][0][1] , "object", "node should have attributes");
        q.equal(ast[1]["stmt"][0][1]["id"], "A", "node id should be A");
        q.ok(ast[1]["stmt"][0][1]["attributes"] instanceof Array, "node should have attributes");

        // for node A's attribute
        q.equal(ast[1]["stmt"][0][1]["attributes"][0], "attributes", "it should be attribute token");
        q.deepEqual(ast[1]["stmt"][0][1]["attributes"][1], { "label": "aaa", "foo": "bar" }, "attirbute values ok");

        // for node B
        q.equal(ast[1]["stmt"][1][0], "node", "first statement should be node");
        q.equal(typeof ast[1]["stmt"][1][1] , "object", "node should have attributes");
        q.equal(ast[1]["stmt"][1][1]["id"], "B", "node id should be B");
        q.ok(ast[1]["stmt"][1][1]["attributes"] instanceof Array, "node should have attributes");

        // for node B's attribute
        q.equal(ast[1]["stmt"][1][1]["attributes"][0], "attributes", "it should be attribute token");
        q.deepEqual(ast[1]["stmt"][1][1]["attributes"][1], { "label": "あいうえお" }, "attirbute values ok");

        // for node C
        q.equal(ast[1]["stmt"][2][0], "node", "node ok");
        q.deepEqual(ast[1]["stmt"][2][1], { "id": "C", "attributes": null }, "node attributes OK ");

        // for edge A -> B
        q.equal(ast[1]["stmt"][3][0], "edge", "edge ok");
        q.equal(ast[1]["stmt"][3][1], "normal", "edge ok");
        q.equal(ast[1]["stmt"][3][2], "A", "edge left side OK");
        q.equal(ast[1]["stmt"][3][3], "B", "edge right side OK");
        q.deepEqual(ast[1]["stmt"][3][4], ["attributes", { "label": "A to B" }], "edge has attributes");

        // for edge B --> A
        q.equal(ast[1]["stmt"][4][0], "edge", "edge ok");
        q.equal(ast[1]["stmt"][4][1], "dotted", "edge ok");
        q.equal(ast[1]["stmt"][4][2], "B", "edge left side OK");
        q.equal(ast[1]["stmt"][4][3], "A", "edge right side OK");
        q.equal(ast[1]["stmt"][4][4], null, "edge has not attributes ");
    });

    q.test('under_score case', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" + 
            "under_score;\n" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.ok(ast[1] instanceof Object, "graph has token attribute");
        q.equal(ast[1]["id"], "seqdiag", "graph attribute should have 'id'");

        q.ok(ast[1]["stmt"] instanceof Array, "graph attribute should have stmt");

        // for node under_score
        q.equal(ast[1]["stmt"][0][0], "node", "first statement should be node");
        q.equal(typeof ast[1]["stmt"][0][1] , "object", "node should have attributes");
        q.equal(ast[1]["stmt"][0][1]["id"], "under_score", "node id should be under_score");

    });

    q.test('comment statement', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" + 
            "// this is comment.\n" +
            "under_score;\n" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.ok(ast[1] instanceof Object, "graph has token attribute");
        q.equal(ast[1]["id"], "seqdiag", "graph attribute should have 'id'");

        q.ok(ast[1]["stmt"] instanceof Array, "graph attribute should have stmt");

        // for node under_score
        q.equal(ast[1]["stmt"][0][0], "comment", "first statement should be comment");
        q.equal(ast[1]["stmt"][0][1] , "this is comment.", "comment OK");

    });

    q.test('edge types', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" + 
            "A -> B;" +
            "B --> C;" +
            "B <-- C;" +
            "A <- B;" +
            "A ->> B;" +
            "B -->> C;" +
            "B <<-- C;" +
            "A <<- B;" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.ok(ast[1] instanceof Object, "graph has token attribute");
        q.equal(ast[1]["id"], "seqdiag", "graph attribute should have 'id'");

        q.ok(ast[1]["stmt"] instanceof Array, "graph attribute should have stmt");

        // for node under_score
        q.equal(ast[1]["stmt"][0][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][0][1] , "normal", "edge type should be normal");

        q.equal(ast[1]["stmt"][1][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][1][1] , "dotted", "edge type should be dotted");

        q.equal(ast[1]["stmt"][2][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][2][1] , "return_dotted", "edge type should be dotted");

        q.equal(ast[1]["stmt"][3][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][3][1] , "return", "edge type should be dotted");

        q.equal(ast[1]["stmt"][4][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][4][1] , "async", "edge type should be dotted");

        q.equal(ast[1]["stmt"][5][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][5][1] , "async_dotted", "edge type should be dotted");

        q.equal(ast[1]["stmt"][6][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][6][1] , "return_async_dotted", "edge type should be dotted");

        q.equal(ast[1]["stmt"][7][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][7][1] , "return_async", "edge type should be dotted");
    });

    q.test('auto return edge', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" + 
            "A => B;" +
            "A ==> B;" +
            "A =>> B;" +
            "A ==>> B;" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.ok(ast[1] instanceof Object, "graph has token attribute");
        q.equal(ast[1]["id"], "seqdiag", "graph attribute should have 'id'");

        q.ok(ast[1]["stmt"] instanceof Array, "graph attribute should have stmt");

        // for node under_score
        q.equal(ast[1]["stmt"][0][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][0][1] , "bidirectional", "edge type should be bidirectional");

        q.equal(ast[1]["stmt"][1][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][1][1] , "bidirectional_dotted", "edge type should be bidirectional_dotted");

        q.equal(ast[1]["stmt"][2][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][2][1] , "bidirectional_async", "edge type should be bidirectional_async");

        q.equal(ast[1]["stmt"][3][0], "edge", "statement should be edge");
        q.equal(ast[1]["stmt"][3][1] , "bidirectional_async_dotted", "edge type should be bidirectional_async_dotted");
    });

    q.test('separator', function() {
        var ast = Seqdiag.Parser.parse("seqdiag {\n" +
            "A -> B;" +
            "=== Separator line ===" +
            "A -> B;" +
            "... Separator line ..." +
            "A -> B;" +
        "}");

        q.ok(ast instanceof Array, "should return tuple");
        q.equal(ast[0], "graph", "token should be graph");

        q.equal(ast[1]["stmt"][0][0], "edge", "statement should be edge");

        q.equal(ast[1]["stmt"][1][0], "separator", "statement should be separator");
        q.equal(ast[1]["stmt"][1][1], "normal", "separator type should be normal");
        q.equal(ast[1]["stmt"][1][2], "Separator line ", "separator comment OK");

        q.equal(ast[1]["stmt"][2][0], "edge", "statement should be edge");

        q.equal(ast[1]["stmt"][3][0], "separator", "statement should be separator");
        q.equal(ast[1]["stmt"][3][1], "delay", "separator type should be normal");
        q.equal(ast[1]["stmt"][3][2], "Separator line ", "separator comment OK");

        q.equal(ast[1]["stmt"][4][0], "edge", "statement should be edge");

    });
})();



test/diagram.js
require('./test_helper.js');
(function() {
    "use strict";

    var q = QUnit;

    q.test("Node", function() {
        var nodeA = new Seqdiag.Node("A");
        q.ok(nodeA instanceof Seqdiag.Node, "instance OK");
        q.equal(nodeA.id, "A", "A OK");
        var nodeB = new Seqdiag.Node("B");
        q.ok(nodeB instanceof Seqdiag.Node, "instance OK");
        q.equal(nodeB.id, "B", "B OK");
        q.equal(nodeA.id, "A", "A OK");
    });

    q.test("Edge", function() {
        var nodeA = new Seqdiag.Node("A");
        var nodeB = new Seqdiag.Node("B");
        var edgeAB = new Seqdiag.Edge(nodeA, nodeB, { label: "a" });
        q.ok(edgeAB instanceof Seqdiag.Edge, "instance OK");
        q.strictEqual(edgeAB.from, nodeA, "from OK");
        q.strictEqual(edgeAB.to, nodeB, "to OK");
    });

    q.test("Diagram", function() {
        var diagram = new Seqdiag.Diagram();
        q.ok(diagram instanceof Seqdiag.Diagram, "instance OK");

        var nodeA = new Seqdiag.Node("A");
        diagram.addNode(nodeA);
        q.equal(diagram.nodes().length, 1, "node A should be added");
        q.strictEqual(diagram.getNodeById("A"), nodeA, "getNodeById should return nodeA");

        var nodeB = new Seqdiag.Node("B");
        diagram.addNode(nodeB);
        q.equal(diagram.nodes().length, 2, "noe B should be added");
        q.strictEqual(diagram.getNodeById("B"), nodeB, "getNodeById should return nodeB");

        var edgeAB = new Seqdiag.Edge(nodeA, nodeB);
        diagram.addEdge(edgeAB);
        q.equal(diagram.edges.length, 1, "edgeAB should be added");

    });

})();


lib/seqdiag.js

/* SYNOPSIS
//
//      var Seqdiag = require('seqdiag');
//      var diagram = new Seqdiag.Diagram();
//      var nodeA = new Seqdiag.Node("A");
//      diagram.addNode(nodeA);
//      var nodeB = new Seqdiag.Node("B");
//      diagram.addNode(nodeB);
//      var edgeAB = new Seqdiag.Edge(nodeA, nodeB);
//      diagram.addEdge(edgeAB);
//
//  If you load it from browser, Seqdiag namespace is exported.
//
// @license
// (c) 2012 Keiji Yoshimi
// This project is released under MIT license.
// See followings: http://www.opensource.org/licenses/MIT
*/
(function(exports)
{
    "use strict";

    // -----------------------------------------------------------------------------------------
    // @class Node
    //
    var Node = function(id, attributes) {
        this.id = id;
        this.attributes = attributes;
        return this;
    };

    // -----------------------------------------------------------------------------------------
    // @class Edge
    //
    var Edge = function(from, to, attributes) {
        this.from = from;
        this.to   = to;
        this.attributes = attributes;
        return this;
    };

    // -----------------------------------------------------------------------------------------
    // @class Separator
    //
    var Separator = function(type, comment) {
        this.type = type;
        this.comment = comment;
        return this;
    };

    // -----------------------------------------------------------------------------------------
    // @class Diagram
    //
    var Diagram = function() {

        this.nodeIds = [];
        this.nodeIdOf = {};
        this.edges = [];
        this.separators = [];
        this.sequences = [];

        return this;
    };

    Diagram.prototype.addNode = function(node) {
        this["nodeIds"].push(node.id);
        this["nodeIdOf"][node.id] = node;
    };

    Diagram.prototype.nodes = function(node) {
        var self = this;
        return this["nodeIds"].map (function(item) {
            return self.getNodeById(item);
        });
    };

    Diagram.prototype.getNodeById = function(nodeId) {
        return this["nodeIdOf"][nodeId];
    };

    Diagram.prototype.addEdge = function(edge) {
        this["edges"].push(edge);
        this["sequences"].push(edge);
    };

    Diagram.prototype.addSeparator = function(separator) {
        this["separators"].push(separator);
        this["sequences"].push(separator);
    };

    // -----------------------------------------------------------------------------------------
    /*
     *  @class DiagramBuilder
     *
     *  build up Diagram object from parsed tokens.
     *
     *  SYNOPSIS:
     *      var tokens = SeqdiagParser.parse(str);
     *      var diagram = DiagramBuilder.build(tokens);
     *      assert(diagram instanceof Seqdiag.Diagram);
     */
    var DiagramBuilder = {
        build: function(tokens) {
            if ( tokens[0] == "graph" ) {
                return this.buildDiagram(tokens[1]);
            }
        },
        buildDiagram: function(tokens) {
            var diagram = new Diagram();
            for (var i = 0; i < tokens.stmt.length; i++ ) {
                switch ( tokens.stmt[i][0] ) {
                    case "node":
                        var node = this.buildNode(diagram, tokens.stmt[i][1]);
                        diagram.addNode(node);
                        break;
                    case "edge":
                        if ( tokens.stmt[i][1].indexOf("bidirectional") > -1 ) {
                            var edges = this.buildBidirectionalEdge(diagram, tokens.stmt[i]);
                            diagram.addEdge(edges[0]);
                            diagram.addEdge(edges[1]);
                        } else {
                            var edge = this.buildEdge(diagram, tokens.stmt[i]);
                            diagram.addEdge(edge);
                        }
                        break;
                    case "separator":
                        var separator = this.buildSeparator(diagram, tokens.stmt[i]);
                        diagram.addSeparator(separator);
                        break;
                }
            }
            return diagram;
        },
        buildNode: function(diagram, attribute) {
            var attrs = attribute["attributes"];
            if ( attrs ) {
                attrs = attrs[1];
            } else {
                attrs = null;
            }
            var node = new Node(attribute["id"], attrs);
            return node;
        },
        buildBidirectionalEdge: function(diagram, attribute){
            var type = attribute[1];
            switch ( type ) {
                case "bidirectional":
                    type = "normal";
                    break;
                case "bidirectional_dotted":
                    type = "dotted";
                    break;
                case "bidirectional_async":
                    type = "async";
                    break;
                case "bidirectional_async_dotted":
                    type = "async_dotted";
                    break;
                default:
                    type = "normal";
                    break;
            }
            var toAttr = {};
            if ( attribute[4] ) {
                toAttr["label"] = attribute[4][1]["return"];
            }
            var fromEdge = this.buildEdge(diagram, [attribute[0], type, attribute[2], attribute[3], attribute[4]]);
            var toEdge = this.buildEdge(diagram, [attribute[0],   type, attribute[3], attribute[2], ["attributes", toAttr]]);
            return [fromEdge, toEdge];
        },
        buildEdge: function(diagram, attribute) {
            var from = diagram.getNodeById(attribute[2]);
            var type = attribute[1];
            if ( ! from ) {
                from = new Node(attribute[2]);
                diagram.addNode(from);
            }
            var to   = diagram.getNodeById(attribute[3]);
            if ( ! to ) {
                to = new Node(attribute[3]);
                diagram.addNode(to);
            }
            var attrs = attribute[4];
            if ( attrs ) {
                attrs = attrs[1];
            } else {
                attrs = {};
            }

            if ( from.id == to.id ) {
                type = "self_reference";
            }
            attrs["type"] = type;
            if ( type.indexOf("return") > -1 ) {
                attrs["isReturn"] = true;
            } else {
                attrs["isReturn"] = false;
            }
            if ( type.indexOf("async") > -1 ) {
                attrs["isAsync"] = true;
            } else {
                attrs["isAsync"] = false;
            }
            if ( type.indexOf("dotted") > -1 ) {
                attrs["isDotted"] = true;
            } else {
                attrs["isDotted"] = false;
            }

            var edge;
            if ( attrs["isReturn"] ) {
                edge = new Edge(to, from, attrs);
            } else {
                edge = new Edge(from, to, attrs);
            }
            return edge;
        },
        buildSeparator: function(diagram, attribute) {
            var sep = new Separator(attribute[1], attribute[2]);
            return sep;
        }
    };

    // -----------------------------------------------------------------------------------------
    // @class Drawer.SVG
    //
    var Drawer = {
        SVG: function(diagram, svg, document) {
            this.diagram = diagram;
            this.document = document;
            this.svg = svg;
            return this;
        }
    };

    Drawer.SVG.prototype.SVG_XML_NAMESPACE = "http://www.w3.org/2000/svg";

    Drawer.SVG.prototype.createSVGElement = function(tagName) {
        return this.document.createElementNS(this.SVG_XML_NAMESPACE, tagName);
    };

    Drawer.SVG.prototype.getNodeRect = function(nodeId) {
        return this.svg.getElementById('node-rect-' + nodeId);
    };

    Drawer.SVG.prototype.draw = function() {
        this.drawNodes();
        this.drawEdges();
        this.drawLifeLines();
    };

    Drawer.SVG.prototype.marginTop = 10;

    Drawer.SVG.prototype.drawNodes = function() {
        var nodes = this.diagram.nodes();
        var metrics = this.calcNodeMetrics();
        var x = metrics["widthMargin"];
        var y = this.marginTop;
        for (var i = 0; i < nodes.length; i++ ) {
            this.drawNode(nodes[i], {
                "x": x,
                "y": y,
                "width": metrics["width"],
                "height": metrics["height"],
            });
            x += metrics["width"] + metrics["widthMargin"];
        }
    };

    Drawer.SVG.prototype.defaultDiagramWidth = 1024;
    Drawer.SVG.prototype.defaultDiagramHeight = 600;
    Drawer.SVG.prototype.defaultWidthMargin = 50;
    Drawer.SVG.prototype.defaultHeightMargin = 30;
    Drawer.SVG.prototype.nodeHeight = 30;

    Drawer.SVG.prototype.calcNodeMetrics = function() {
        var svgWidth = parseInt(this.svg.getAttribute("width")) || this.defaultDiagramWidth;
        var svgHeight = parseInt(this.svg.getAttribute("height")) || this.defaultDiagramHeight;
        var widthMargin = this.defaultWidthMargin;
        var heightMargin = this.defaultHeightMargin;
        // svgWidth = ( nodeWidth * length ) + ( widthMargin * ( length + 1 ) )
        var nodeLength = this.diagram.nodes().length;
        var nodeWidth = ( ( svgWidth - widthMargin * ( nodeLength + 1 ) ) / nodeLength );
        var nodeHeight = this.nodeHeight;
        return {
            "width": nodeWidth,
            "widthMargin": widthMargin,
            "height": nodeHeight,
            "heightMargin": heightMargin
        }
    };

    Drawer.SVG.prototype.defaultBackgroundColor = "#f9f9ff"; // white (little blue)
    Drawer.SVG.prototype.defaultStrokeColor = "#000033"; // black (little navy)

    Drawer.SVG.prototype.getNodeFillColor = function(node) {
        return Drawer.SVG.prototype.defaultBackgroundColor;
    };

    Drawer.SVG.prototype.getNodeStrokeColor = function(node) {
        return Drawer.SVG.prototype.defaultStrokeColor;
    };

    Drawer.SVG.prototype.charWidth = 8;
    Drawer.SVG.prototype.charHeight = 8;

    Drawer.SVG.prototype.drawNode = function(node, params) {
        var rect = this.createSVGElement('rect');
        rect.setAttribute("id", "node-rect-" + node.id);
        rect.setAttribute("class", "node-rect");
        rect.setAttribute("width", params["width"]);
        rect.setAttribute("height", params["height"]);
        rect.setAttribute("x", params["x"]);
        rect.setAttribute("y", params["y"]);
        rect.setAttribute("fill", this.getNodeFillColor(node));
        rect.setAttribute("stroke", this.getNodeStrokeColor(node));

        this.svg.appendChild(rect);

        var textbox = this.createSVGElement('text');
        if ( node.attributes && node.attributes["label"] !== undefined ) {
            textbox.textContent = node.attributes["label"];
        } else {
            textbox.textContent = node.id;
        }
        textbox.setAttribute("id", "node-textbox-" + node.id);
        textbox.setAttribute("class", "node-text");
        textbox.setAttribute("x", params["x"] + params["width"] / 2.0 - textbox.textContent.length * this.charWidth / 2.0);
        textbox.setAttribute("y", params["y"] + params["height"] / 2.0 + ( this.charHeight / 2.0 ));

        this.svg.appendChild(textbox);
    };

    Drawer.SVG.prototype.drawEdges = function() {
        var y = this.marginTop + this.nodeHeight;
        var svgHeight = parseInt(this.svg.getAttribute("height")) || this.defaultDiagramHeight;
        var edgeMargin = ( svgHeight - y ) / ( this.diagram.edges.length + 1 );
        y += edgeMargin;
        var edges = this.diagram.edges;
        for (var i = 0; i < edges.length; i++ ) {
            if ( edges[i].attributes["type"] == "self_reference" ) {
                this.drawSelfReferenceEdge(edges[i], {
                    "y": y,
                });
            } else {
                this.drawEdge(edges[i], {
                    "y": y,
                });
            }
            y += edgeMargin;
        }
    };

    Drawer.SVG.prototype.getEdgeStrokeColor = function(edge) {
        return this.defaultStrokeColor;
    };

    Drawer.SVG.prototype.drawEdge = function(edge, params) {
        var path = this.createSVGElement("path");
        var from = this.getNodeRect(edge.from.id);
        var to   = this.getNodeRect(edge.to.id);

        var y = params["y"];
        var fromX   = parseInt(from.getAttribute("x")) + parseInt(from.getAttribute("width")) / 2.0;
        var toX     =   parseInt(to.getAttribute("x")) + parseInt(to.getAttribute("width")) / 2.0;
        path.setAttribute("d", "M " + fromX + " " + y + " L " + toX + " " + y);
        path.setAttribute("stroke", this.getEdgeStrokeColor(edge));
        if ( edge.attributes["isDotted"] ) {
            path.setAttribute("stroke-dasharray", 5);
        }

        this.drawEdgeArrowHead(edge, {
            "fromX": fromX,
            "toX"  : toX,
            "y"    : y
        });

        this.svg.appendChild(path);

        if ( edge.attributes && edge.attributes["label"] !== undefined ) {
            this.drawEdgeText(edge, {
                "fromX": fromX,
                "toX"  : toX,
                "y"    : y
            });
        }
    };

    Drawer.SVG.prototype.drawSelfReferenceEdge = function(edge, params)
    {
        var y = params["y"];

        var target = this.getNodeRect(edge.from.id);
        var nextNode = null;
        var nodes = this.diagram.nodes();
        for ( var i = 0; i < nodes.length; i++ ) {
            if ( target.id == nodes[i].id ) {
                nextNode = nodes[i+1];
                break;
            }
        }
        // for right node's self-reference.
        if ( !nextNode ) {
            nextNode = nodes[nodes.length-1-1];
        }
        var nextNodeRect = this.getNodeRect(nextNode.id);
        var fromX   = parseInt(target.getAttribute("x")) + parseInt(target.getAttribute("width")) / 2.0;
        var toX   = parseInt(nextNodeRect.getAttribute("x")) + parseInt(nextNodeRect.getAttribute("width")) / 2.0;
        if ( fromX < toX ) {
            toX = fromX + ( toX - fromX ) / 2;
        } else {
            toX = fromX + ( fromX - toX ) / 2;
        }

        var fromPath = this.createSVGElement("path");
        fromPath.setAttribute("stroke", this.defaultStrokeColor);
        fromPath.setAttribute("d", "M " + fromX + " " + y + " L " + toX + " " + y);
        var toPath = this.createSVGElement("path");
        var selfReferenceEdgeMargin = 15;
        toPath.setAttribute("stroke", this.defaultStrokeColor);
        toPath.setAttribute("d", "M " + toX + " " + ( y + selfReferenceEdgeMargin ) + " L " + fromX + " " + ( y + selfReferenceEdgeMargin ) );

        var sidePath = this.createSVGElement("path");
        sidePath.setAttribute("stroke", this.defaultStrokeColor);
        sidePath.setAttribute("d", "M " + toX + " " + y + " L " + toX + " " + ( y + selfReferenceEdgeMargin ));

        this.svg.appendChild(fromPath);
        this.svg.appendChild(toPath);
        this.svg.appendChild(sidePath);

        this.drawEdgeArrowHead(edge, {
            "fromX": toX,
            "toX": fromX,
            "y" : y + selfReferenceEdgeMargin
        });

        this.drawEdgeText(edge, {
            "fromX": fromX,
            "toX": toX,
            "y" : y
        });

    };

    Drawer.SVG.prototype.arrowWidth = 10;
    Drawer.SVG.prototype.arrowHeight = 6;

    Drawer.SVG.prototype.drawEdgeArrowHead = function(edge, options) {
        var fromX = options["fromX"];
        var toX   = options["toX"];
        var y     = options["y"];

        var arrowWidth = this.arrowWidth;
        var arrowHeight = this.arrowHeight;

        var arrowFromX = null;
        if ( fromX < toX ) {
            arrowFromX = toX - arrowWidth;
        } else {
            arrowFromX = toX + arrowWidth;
        }
        if ( edge.attributes["isAsync"] ) {
            var arrowUpper = this.createSVGElement("path");
            arrowUpper.setAttribute("stroke", this.getEdgeStrokeColor(edge));
            arrowUpper.setAttribute("d", "M " + arrowFromX + " " + ( y - arrowHeight / 2 ) + " L " + toX + " " + y);
            this.svg.appendChild(arrowUpper);

            var arrowSupper = this.createSVGElement("path");
            arrowSupper.setAttribute("stroke", this.getEdgeStrokeColor(edge));
            arrowSupper.setAttribute("d", "M " + arrowFromX + " " + ( y + arrowHeight / 2 ) + " L " + toX + " " + y);
            this.svg.appendChild(arrowSupper);

        } else {
            var arrowHead = this.createSVGElement("polygon");
            arrowHead.setAttribute("stroke", this.getEdgeStrokeColor(edge));
            arrowHead.setAttribute("points", [
                [arrowFromX, y - arrowHeight/2],
                [arrowFromX, y + arrowHeight/2],
                [toX, y]
            ].map(function(i) { return i.join(",") }).join(" "));
            this.svg.appendChild(arrowHead);
        }
    };

    Drawer.SVG.prototype.drawEdgeText = function (edge, options)
    {
        var fromX = options["fromX"];
        var toX   = options["toX"];
        var y     = options["y"];

        var textbox = this.createSVGElement('text');
        textbox.textContent = edge.attributes["label"];
        textbox.setAttribute("class", "edge-text");
        if ( fromX < toX ) {
            textbox.setAttribute("x", fromX + 5);
        } else {
            textbox.setAttribute("x", toX + 20);
        }
        textbox.setAttribute("y", y - 5);
        this.svg.appendChild(textbox);
    };

    Drawer.SVG.prototype.drawLifeLines = function() {
        var nodes = this.diagram.nodes();
        for (var i = 0; i < nodes.length; i++) {
            var node = nodes[i];
            this.drawLifeLine(nodes[i]);
        }
    };

    Drawer.SVG.prototype.drawLifeLine = function(node) {
        var rect = this.getNodeRect(node.id);
        var rectCenterX = parseInt(rect.getAttribute("x")) + parseInt(rect.getAttribute("width")) / 2.0;
        var rectBottomY = parseInt(rect.getAttribute("y")) + parseInt(rect.getAttribute("height"));

        var line = this.createSVGElement('path');
        line.setAttribute("d", "M " + rectCenterX + " " + rectBottomY + "L " + rectCenterX + " " + this.defaultDiagramHeight);
        line.setAttribute("stroke", "#000000");
        line.setAttribute("stroke-dasharray", 5);
        this.svg.appendChild(line);
    };

    // -----------------------------------------------------------------------------------------
    // declare exports followings:
    exports.Node = Node;
    exports.Edge = Edge;
    exports.Separator = Separator;
    exports.Diagram = Diagram;
    exports.DiagramBuilder = DiagramBuilder;
    exports.Drawer = Drawer;

})(typeof exports === 'undefined' ? this['Seqdiag'] = {} : exports );
